File: src/main/java/de/lmu/bio/calcium/io/CaNIXImporter.java
package de.lmu.bio.calcium.io;


import de.lmu.bio.calcium.model.CaImage;
import de.lmu.bio.calcium.model.CaNeuron;
import de.lmu.bio.calcium.model.CaRoiBox;
import de.lmu.bio.calcium.model.CaRoiClass;
import ij.gui.PolygonRoi;
import ij.gui.Roi;
import org.g_node.nix.*;
import ucar.ma2.Array;

import java.awt.*;
import java.util.Comparator;
import java.util.List;

public class CaNIXImporter extends CaImporter {

    private CaNeuron neuron = null;

    private String path;

    //progress tracking
    private int imagesTotal = 0;
    private int imagesProcessed = 0;

    public CaNIXImporter(String path) {
        super("Importing");
        this.path = path;
    }

    @Override
    public CaNeuron getNeuron() {
        return neuron;
    }

    @Override
    protected void runTask() throws Exception {
        File fd = File.open(path, FileMode.ReadOnly);

        List<Block> blocks = fd.getBlocks(b -> b.getType().equals("neuron"));
        if (blocks.size() != 1) {
            throw new RuntimeException("None or too many blocks with type neuron found");
        }

        Block block = blocks.get(0);
        Section meta = block.getMetadata();

        neuron = new CaNeuron(block.getName());

        if (meta.hasProperty("region")) {
            neuron.setRegion(meta.getProperty("region").getValues().get(0).getString());
        }

        if (meta.hasProperty("age")) {
            neuron.setAge(meta.getProperty("age").getValues().get(0).getString());
        }

        if (meta.hasProperty("condition")) {
            neuron.setCondition(meta.getProperty("condition").getValues().get(0).getString());
        }

        if (meta.hasProperty("sex")) {
            neuron.setSex(meta.getProperty("sex").getValues().get(0).getString());
        }

        if (meta.hasProperty("subregion")) {
            neuron.setSubregion(meta.getProperty("subregion").getValues().get(0).getString());
        }

        if (meta.hasProperty("litter")) {
            neuron.setLitter(meta.getProperty("litter").getValues().get(0).getString());
        }

        List<Group> imgGroups = block.getGroups(g -> g.getType().equals("image.ca"));
        imagesTotal = imgGroups.size();

        imgGroups.stream().map(this::importImage)
                .sorted(Comparator.comparingInt(i -> Integer.parseInt(i.getName())))
                .forEach(i -> neuron.add(i));

        fd.close();
    }

    private CaImage importImage(Group g) {
        fireTaskProgress(imagesProcessed, imagesTotal, "Importing " + g.getName());
        Section meta = g.getMetadata();

        String filename = meta.getProperty("filename").getValues().get(0).getString();
        java.io.File ps = new java.io.File(new java.io.File(path).getParentFile(), filename);

        String filepath;
        if (ps.exists()) {
            filepath = ps.getAbsolutePath();
        } else {
            filepath = meta.getProperty("original_path").getValues().get(0).getString();
        }

        CaImage img = new CaImage(filepath);

        CaImage.Metadata im = img.new Metadata();

        im.ctime = meta.getProperty("creation_time").getValues().get(0).getLong();
        im.channels = meta.getProperty("channels").getValues().get(0).getInt();

        if (meta.hasProperty("condition")) {
            img.setCondition(meta.getProperty("condition").getValues().get(0).getString());
        }

        DataArray chans = g.getDataArray(g.getName() + ".channels");
        RangeDimension dim = chans.getDimensions().get(0).asRangeDimension();

        int extent = chans.getDataExtent().getData()[0];
        int[] data = new int[extent];
        chans.getData(data, new NDSize(1, extent), new NDSize(1, 0)); //FIXME check
        double[] ticks = dim.getTicks();

        im.planeInfo = new CaImage.PlaneInfo[data.length];
        for (int i = 0; i < data.length; i++) {
            im.planeInfo[i] = img.new PlaneInfo();
            im.planeInfo[i].timePoint = i;
            im.planeInfo[i].channel = data[i];
            im.planeInfo[i].deltaT = ticks[i];
        }
        im.timePoints = data.length;
        img.setMetadata(im);

        List<DataArray> roiData = g.getDataArrays(d -> d.getType().startsWith("roi.pt"));

        roiData.stream().map(d -> {

            Section rm = d.getMetadata();
            int roiType = rm.getProperty("type").getValues().get(0).getInt();

            Roi roi = null;
            if (roiType == Roi.RECTANGLE) {
                roi = importRectRoi(d);
            } else if (roiType == Roi.POLYLINE || roiType == Roi.FREELINE) {
                roi = importPolyRoi(d);
            } else {
                return null;
            }

            String[] comps = d.getName().split("\\.");
            roi.setName(comps[comps.length - 1].toUpperCase());

            CaRoiBox box =  new CaRoiBox(roi);
            Color color = null;
            if (rm.hasProperty("color")) {
                int rgb = rm.getProperty("color").getValues().get(0).getInt();
                color = new Color(rgb);
            } else {
                CaRoiClass cls = box.getRoiClass();
                color = cls.roiColor();
            }

            roi.setStrokeColor(color);
            return box;
        }).filter(b -> b != null).forEach(img::add);

        imagesProcessed++;
        return img;
    }

    Roi importPolyRoi(DataArray d) {
        Section rm = d.getMetadata();
        int roiType = rm.getProperty("type").getValues().get(0).getInt();
        double strokeWidth = rm.getProperty("strokeWidth").getValues().get(0).getDouble();

        NDSize shape = d.getDataExtent();
        float[] xy = new float[(int) shape.getElementsProduct()];
        d.getData(xy, shape, new NDSize(2, 0));

        int n = shape.getData()[1];
        float[] x = new float[n];
        float[] y = new float[n];

        System.arraycopy(xy, 0, x, 0, n);
        System.arraycopy(xy, n, y, 0, n);
        Roi roi =  new PolygonRoi(x, y, roiType);
        if (strokeWidth > 1.0) {
            roi.setStrokeWidth(strokeWidth);
        }
        return roi;
    }

    Roi importRectRoi(DataArray d) {
        float[] pts = new float[4];
        d.getData(pts, new NDSize(1, 4), new NDSize(1, 0));

        return new Roi(pts[0], pts[1], pts[2], pts[3]);
    }

}

File: src/main/java/de/lmu/bio/calcium/io/CaNIXImporter.kt
package de.lmu.bio.calcium.io

import de.lmu.bio.calcium.CaTask
import de.lmu.bio.calcium.model.CaImage
import de.lmu.bio.calcium.model.CaNeuron
import de.lmu.bio.calcium.model.CaRoiBox
import ij.IJ
import ij.gui.PolygonRoi
import ij.gui.Roi
import org.g_node.nix.*;

class CaNIXImporter(val path: String)
  : CaImporter("Importing") {

  var imagesTotal = 0
  var imagesProcessed = 0

  // importer state
  var theNeuron: CaNeuron? = null

  override fun runTask() {
    val fd = File.open(path, FileMode.ReadOnly)
    val block = fd.getBlocks{ b -> b.type == "neuron" }.lastOrNull() ?: throw RuntimeException("No neuron in file")
    val meta = block.metadata

    theNeuron = CaNeuron(block.name)

    if (meta.hasProperty("region")) {
      theNeuron?.region = meta.getProperty("region").values[0].string
    }

    if (meta.hasProperty("age")) {
      theNeuron?.age = meta.getProperty("age").values[0].string
    }

    if (meta.hasProperty("condition")) {
      theNeuron?.condition = meta.getProperty("condition").values[0].string
    }

    if (meta.hasProperty("sex")) {
      theNeuron?.sex = meta.getProperty("sex").values[0].string
    }

    if (meta.hasProperty("subregion")) {
      theNeuron?.subregion = meta.getProperty("subregion").values[0].string
    }

    if (meta.hasProperty("litter")) {
      theNeuron?.comment = meta.getProperty("litter").values[0].string
    }

    var imageGroups = block.getGroups { g -> g.type == "image.ca"}
    imagesTotal = imageGroups.count()
    
    imageGroups.map { g ->
      importImage(g)
    }.sortedBy {
      i -> i.name.toInt()
    }.forEach { i ->
      theNeuron?.add(i)
    }

    imageGroups.forEach { g -> g.close() }
    block.close()
    fd.close()
  }

  fun importImage(parent: Group) : CaImage {
    fireTaskProgress(imagesProcessed, imagesTotal, "Importing " + parent.name)
    val meta = parent.metadata

    val filename = meta.getProperty("filename").values[0].string
    var ps = java.io.File(java.io.File(path).parentFile, filename)

    var path = if (ps.exists()) {
      ps.absolutePath
    } else {
      meta.getProperty("original_path").values[0].string
    }

    var img = CaImage(path)
    var md = img.Metadata()

    //we should check that we actually have that info
    md.ctime = meta.getProperty("creation_time").values[0].long
    md.channels = meta.getProperty("channels").values[0].int

    val chans = parent.getDataArray(parent.name + ".channels")
    val data = IntArray(chans.dataExtent.data[0])
    val dim = chans.dimensions[0].asRangeDimension()
    val ticks = dim.ticks

    md.planeInfo = Array<CaImage.PlaneInfo>(data.size, { idx ->
      var info = img.PlaneInfo()
      info.timePoint = idx
      info.channel = data[idx]
      info.deltaT = ticks[idx]
      info
    })

    img.metadata = md
    img.metadata.timePoints = md.planeInfo.size

    var roiData = parent.getDataArrays {
      d -> d.type.startsWith("roi.pt", true)
    }

    roiData.map { d ->
      val shape = d.dataExtent
      val xy = FloatArray(shape.elementsProduct.toInt())
      d.getData(xy, shape, NDSize(2, 0))
      val roiMeta = d.metadata
      val roiType = roiMeta.getProperty("type").values[0].int
      val strokeWidth = roiMeta.getProperty("strokeWidth").values[0].double

      val x = xy.copyOfRange(0, shape.data[1])
      val y = xy.copyOfRange(shape.data[1], xy.size)
      var roi = PolygonRoi(x, y, roiType)
      roi.strokeWidth = strokeWidth.toFloat()
      roi.name = d.name.split(".").last().toUpperCase()

      CaRoiBox(roi)
    }.forEach { roi -> img.add(roi) }
    meta.close()
    chans.close()

    imagesProcessed++
    return img
  }

  override fun getNeuron(): CaNeuron? {
    return theNeuron
  }
}